clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2220,23 +2220,103 @@ namespace {
 class ExtractTypeForDeductionGuide
   : public TreeTransform<ExtractTypeForDeductionGuide> {
   llvm::SmallVectorImpl<TypedefNameDecl *> &MaterializedTypedefs;
+  ClassTemplateDecl *NestedPattern;
+  const MultiLevelTemplateArgumentList *OuterInstantiationArgs;
 
 public:
   typedef TreeTransform<ExtractTypeForDeductionGuide> Base;
   ExtractTypeForDeductionGuide(
       Sema &SemaRef,
-      llvm::SmallVectorImpl<TypedefNameDecl *> &MaterializedTypedefs)
-      : Base(SemaRef), MaterializedTypedefs(MaterializedTypedefs) {}
+      llvm::SmallVectorImpl<TypedefNameDecl *> &MaterializedTypedefs,
+      ClassTemplateDecl *NestedPattern,
+      const MultiLevelTemplateArgumentList *OuterInstantiationArgs)
+      : Base(SemaRef), MaterializedTypedefs(MaterializedTypedefs),
+        NestedPattern(NestedPattern),
+        OuterInstantiationArgs(OuterInstantiationArgs) {}
 
   TypeSourceInfo *transform(TypeSourceInfo *TSI) { return TransformType(TSI); }
 
+  /// Returns true if it's safe to substitute \p Typedef with
+  /// \p OuterInstantiationArgs.
+  bool mightReferToOuterTemplateParameters(TypedefNameDecl *Typedef) {
+    if (!NestedPattern)
+      return false;
+
+    static auto WalkUp = [](DeclContext *DC, DeclContext *TargetDC) {
+      if (DC->Equals(TargetDC))
+        return true;
+      while (DC->isRecord()) {
+        if (DC->Equals(TargetDC))
+          return true;
+        DC = DC->getParent();
+      }
+      return false;
+    };
+
+    if (WalkUp(Typedef->getDeclContext(), NestedPattern->getTemplatedDecl()))
+      return true;
+    if (WalkUp(NestedPattern->getTemplatedDecl(), Typedef->getDeclContext()))
+      return true;
+    return false;
+  }
+
+  QualType
+  RebuildTemplateSpecializationType(TemplateName Template,
+                                    SourceLocation TemplateNameLoc,
+                                    TemplateArgumentListInfo &TemplateArgs) {
+    if (!OuterInstantiationArgs ||
+        !isa_and_present<TypeAliasTemplateDecl>(Template.getAsTemplateDecl()))
+      return Base::RebuildTemplateSpecializationType(Template, TemplateNameLoc,
+                                                     TemplateArgs);
+
+    auto *TATD = cast<TypeAliasTemplateDecl>(Template.getAsTemplateDecl());
+    auto *Pattern = TATD;
+    while (Pattern->getInstantiatedFromMemberTemplate())
+      Pattern = Pattern->getInstantiatedFromMemberTemplate();
+    if (!mightReferToOuterTemplateParameters(Pattern->getTemplatedDecl()))
+      return Base::RebuildTemplateSpecializationType(Template, TemplateNameLoc,
+                                                     TemplateArgs);
+
+    Decl *NewD = SemaRef.SubstDecl(
+        TATD, SemaRef.getASTContext().getTranslationUnitDecl(),
```

Comment:
This is what we've done since D80743, when @zygoloid [suggested](https://reviews.llvm.org/D80743?id=266940#inline-745424) that we delay setting the `DeclContext`s until the `CXXDeductionGuideDecl` is formed. This way, I *guess* (I didn't dig into the details) we could have template parameters attach to the right depths (e.g. presumably 1) rather than the depth relative to the outer class templates (`DC`).

Reply:
I see, thanks for the link.

There is a subtle difference: 
- In `D80743`, we create a `Decl` by calling `TypeAliasDecl::Create(.., TU, ...)`, which only sets the member `TypeAliasDecl::DeclCtx` to `TU`.
- While here, we pass the `TU` to the `SubstDecl`. This has the side effect of adding the `NewD` into the `TU` (in the substitution implementation, `TU->addDecl(NewD)` is called). Consequently, in the AST dump, you will see `NewD` under the `TranslationUnit`, which doesn't look right. I guess we can fix this by calling `TU->removeDecl(NewD)`.


Thinking more about this, we should have this information in the AST.

Given the example:

```cpp
template <class T> struct Outer {
  using Alias = S<T>;
  template <class U> struct Inner {
    Inner(Alias);
  };
};

Outer<int>::Inner inner(S<int>());
```

We're performing a substitution for the `Alias` with `T = int` here. At this point, the `Outer<int>` class specialization should be available. Inside this specialization, we should have the `TypeAliasDecl` `S<int>`. 
So an alternative would be to reuse this `Decl` rather than creating a new one. I'm not sure if this is feasible. It looks like we don't have a good way to get it from the primary `TypeAliasTemplateDecl` (no specializations in the AST node). A hacky way is to perform a lookup from `Outer<int>`.

Reply:
> There is a subtle difference:

Thanks for spotting that! This is indeed something I've overlooked.
Looking at the `SubstDecl` again, I think we can take the middle-ground approach: we call `TemplateDeclInstantiator::InstantiateTypeAliasTemplateDecl()` directly when we want to instantiate a TypedefNameDecl.

This is a plausible approach to me because

1) We could save the recursion into the `DeclVisitor` from `SubstDecl` because we already know the `Decl` is a `TypedefDecl`.

2) This avoids the accidental Decl injection, although we have to tweak the `TemplateDeclInstantiator` a bit.

3) Re: `So an alternative would be to reuse this Decl rather than creating a new one`: granted that we could find that instantiated Decl, we still have to *copy* the Decl, which is in part a duplicate work of the instantiation logic, because we want the transformed Decl to live in the CXXDeductionGuideDecl - I think we can still leverage this instantiation mechanism to keep it intelligible.